Simulator/Utils/CreateFunctionMap.cpp

```diff
@@ -0,0 +1,151 @@
+/**
+ * @file CreateFunctionMap.cpp
+ *
+ * @ingroup Simulator/Utils
+ *
+ * @brief The templated CreateFunctionMap allows each class to create a concrete
+ * instance of itself at runtime.
+ *
+ */
+
+#include <map>
+#include <string>
+
+// Connections
+#include "Connections/NG911/Connections911.h"
+#include "Connections/Neuro/ConnGrowth.h"
+#include "Connections/Neuro/ConnStatic.h"
+
+// Edges
+#include "Edges/NG911/All911Edges.h"
+#include "Edges/Neuro/AllDSSynapses.h"
+#include "Edges/Neuro/AllDynamicSTDPSynapses.h"
+#include "Edges/Neuro/AllSTDPSynapses.h"
+#include "Edges/Neuro/AllSpikingSynapses.h"
+
+// Layout
+#include "Layouts/NG911/Layout911.h"
+#include "Layouts/Neuro/DynamicLayout.h"
+#include "Layouts/Neuro/FixedLayout.h"
+
+// Vertices
+#include "Vertices/NG911/All911Vertices.h"
+#include "Vertices/Neuro/AllIZHNeurons.h"
+#include "Vertices/Neuro/AllLIFNeurons.h"
+
+// Recorder
+#include "Recorders/IRecorder.h"
+#include "Recorders/NG911/Xml911Recorder.h"
+#include "Recorders/Neuro/XmlGrowthRecorder.h"
+#include "Recorders/Neuro/XmlSTDPRecorder.h"
+#include "Recorders/XmlRecorder.h"
+
+#if defined(HDF5)
+   #include "Recorders/Hdf5Recorder.h"
+   #include "Recorders/Neuro/Hdf5GrowthRecorder.h"
+#endif
+
+// MTRand
+#include "RNG/MTRand.h"
+#include "RNG/Norm.h"
+
+/**
+ *
+ * The templated CreateFunctionMap allows each class to create a concrete
+ * instance of itself at runtime. This approach helps to keep the code organized
+ * and avoids circular dependencies that can make the code less error-prone.
+ *
+ * How it avoids circular dependency?
+ * If a map was passed from the base class with all the instance of its derived
+ * class, it would create a circular dependency. Meaning, that the derived class
+ * would need to reference the base class, and the base class would also need to
+ * reference the derived class for the sake of this map, leading to an infinite
+ * loop of references.
+ *
+ * In addition, the templated CreateFunctionMap safeguards against creating a
+ * template of an undesired type. A static assert is used to check for any
+ * undesired type; thereby generating a compile-time error for any request to
+ * instantiate a template for a type that has not been explicitly defined.
+ *
+ * Why the T* argument?
+ * C++ functions resolve polymorphic behavior based on their input arguments,
+ * not their return types. This is because the compiler generates a separate
+ * instance of the function for each function signature. For the templated
+ * CreateFunctionMap, the T* argument is only used to resolve polymorphic
+ * behavior. Also, a pointer is more flexible to pass as a parameter than an
+ * object as the class may or may not have a public constructor and destructor.
```

Comment:
Also, you can create a pointer, `T*`, without having to create an instance of the class `T`. In this case, it is essential, because the function expects a pointer to the (abstract) base class.